### PR TITLE
[#180907761] QueryCleanupManager NPE fix

### PR DESCRIPTION
**Pull request by prajjwolmondal**
[Ticket](https://www.pivotaltracker.com/story/show/180907761)

# Description
Handled an edge case that can arise where `started_at` column isn't populated. Also added a `@Scheduled` method to delete old entries from the `query_job` table.

### CODE REVIEW

File: src/main/resources/application.yml [hunk 1]
```diff
@@ -40,8 +40,12 @@ app:
     oauth-client:
       resource: "${app.indexing-service.base-uri}"
     publisher-catalog-name: publisher_data
-  query-cleanup-interval: "*/10 * * ? * *" # Every 10 seconds
-  query-cleanup-timeout-in-seconds: 120
+  query-cleanup:
+    cron-interval: "*/10 * * ? * *" # Every 10 seconds
+    query-timeout-in-seconds: 120
```
Comment (sharvari-kapadia) [inline src/main/resources/application.yml:45]: Since you have `query-cleanup` as the parent key, how about we use `timeout-in-seconds` as the key here?

File: src/main/resources/application.yml [hunk 2]
```diff
@@ -40,8 +40,12 @@ app:
...
+  query-job-cleanup:
+    cron-interval: "0 0 1 * * ?" # Every day at 1am
+    query-deletion-timeout-in-days: 1
```
Comment (sharvari-kapadia) [inline src/main/resources/application.yml:48]: Same like above, how about we use `timeout-in-seconds` as the key here?

File: src/main/resources/application.yml (hunk 1 above)
Comment (prajjwolmondal) [reply src/main/resources/application.yml]: Done

File: src/main/resources/application.yml (hunk 2 above)
Comment (prajjwolmondal) [reply src/main/resources/application.yml]: Done

File: src/main/java/com/dnastack/ga4gh/dataconnect/adapter/trino/QueryCleanupManager.java [hunk 3]
```diff
@@ -18,23 +18,24 @@
...
-    @Scheduled(cron = "${app.query-cleanup-interval}")
-    public void cleanupOldQueries() {
+    @Scheduled(cron = "${app.query-cleanup.cron-interval}")
+    public void terminateOldQueries() {
         List<QueryJob> queryJobList = jdbi.withExtension(QueryJobDao.class, dao -> {
             Instant oldQueryTimestamp = Instant.now().minusSeconds(queryCleanupTimeoutInSeconds);
             return dao.getOldQueries(oldQueryTimestamp);
         });
-        if (queryJobList.isEmpty()) {
-            log.info("No old queries present, skipping cleanup.");
```
Comment (sharvari-kapadia) [inline src/main/java/com/dnastack/ga4gh/dataconnect/adapter/trino/QueryCleanupManager.java:36]: Was this removed because it was generating lot of unnecessary logs?

File: src/main/java/com/dnastack/ga4gh/dataconnect/repository/QueryJobDao.java [hunk 4]
```diff
@@ -28,9 +28,12 @@
     @SqlUpdate("UPDATE query_job SET last_activity_at = now(), finished_at = now() where id = :id")
     void setQueryFinishedAndLastActivityTime(@Bind String id);
 
-    @SqlQuery("SELECT * FROM query_job WHERE last_activity_at < :lastActivity AND finished_at IS NULL")
+    @SqlQuery("SELECT * FROM query_job WHERE (last_activity_at IS NOT NULL AND started_at IS NOT NULL) AND last_activity_at < :lastActivity AND finished_at IS NULL")
```
Comment (sharvari-kapadia) [inline src/main/java/com/dnastack/ga4gh/dataconnect/repository/QueryJobDao.java:31]: Why did you need to add `started_at IS NOT NULL` ?

File: src/main/resources/application.yml [hunk 5]
```diff
@@ -40,8 +40,12 @@ app:
...
       resource: "${app.indexing-service.base-uri}"
     publisher-catalog-name: publisher_data
-  query-cleanup-interval: "*/10 * * ? * *" # Every 10 seconds
-  query-cleanup-timeout-in-seconds: 120
+  query-cleanup:
+    cron-interval: "*/10 * * ? * *" # Every 10 seconds
+    timeout-in-seconds: 120
+  query-job-cleanup:
+    cron-interval: "0 0 1 * * ?" # Every day at 1am
+    deletion-timeout-in-days: 1
```
Comment (sharvari-kapadia) [inline src/main/resources/application.yml:48]: Why are we deleting entries from the table?
I would recommend keeping the entries at-least for 7 days, just in case we want to debug something in future. Wdyt?

**Review by sharvari-kapadia — APPROVED** [review]
LGTM.
Left some small comments & clarifying questions.

File: src/main/java/com/dnastack/ga4gh/dataconnect/adapter/trino/QueryCleanupManager.java (hunk 3 above)
Comment (prajjwolmondal) [reply src/main/java/com/dnastack/ga4gh/dataconnect/adapter/trino/QueryCleanupManager.java:36]: Yeah, since the method is run every 10s this ends up getting posted a lot

File: src/main/resources/application.yml (hunk 5 above)
Comment (prajjwolmondal) [reply src/main/resources/application.yml:48]: That was [Patricks suggestion](https://www.pivotaltracker.com/story/show/180907761/comments/229007984) as there's no need to store all entries in `query_job`. 

I agree with increasing the deletion timeout to 7 days

File: src/main/java/com/dnastack/ga4gh/dataconnect/repository/QueryJobDao.java (hunk 4 above)
Comment (prajjwolmondal) [reply src/main/java/com/dnastack/ga4gh/dataconnect/repository/QueryJobDao.java:31]: AH my bad, it should be `next_page_url IS NOT NULL` since that's what we use and not `started_at`

File: src/main/java/com/dnastack/ga4gh/dataconnect/repository/QueryJobDao.java [hunk 6]
```diff
@@ -28,9 +28,12 @@
...
+    @SqlQuery("SELECT * FROM query_job WHERE next_page_url IS NOT NULL AND last_activity_at < :lastActivity AND finished_at IS NULL")
     List<QueryJob> getOldQueries(@Bind Instant lastActivity);
 
-    @SqlQuery("SELECT next_page_url FROM query_job WHERE id = :id")
-    String getNextPageUrl(@Bind String id);
+    @SqlQuery("SELECT id FROM query_job WHERE last_activity_at < now()::DATE - :timeoutInDays")
+    List<String> getQueryJobIdsToDelete(@Bind int timeoutInDays);
+
+    @SqlUpdate("DELETE FROM query_job WHERE id IN (<queryJobIds>)")
+    void deleteOldQueryJobs(@BindList List<String> queryJobIds);
```
Comment (prajjwolmondal) [inline src/main/java/com/dnastack/ga4gh/dataconnect/repository/QueryJobDao.java:38]: Combine these two, and return the # of rows deleted
